back-end/java/main/java/src/Filtering/Node_KDTree/KDTreeNode.java

```diff
@@ -0,0 +1,135 @@
+   * point along one of the three dimensions (roomSize, roomCapacity, or roomNumber).
+   */
+  private final DormRoom value;
+
+  /**
+   * The left child node, containing dorm rooms that are "less" than this node's room
+   * along the current splitting dimension.
+   */
+  private final KDTreeNode left;
+
+  /**
+   * The right child node, containing dorm rooms that are "greater" or equal to this node's room
+   * along the current splitting dimension.
+   */
+  private final KDTreeNode right;
+
+  // Since we know the KD-tree is split on three dimensions, we fix K = 3.
+  private static final int K = 3;
+
+
+  /**
+   * Constructs a {@code KDTreeNode} from a list of {@link DormRoom}s by recursively
+   * selecting medians and splitting along the appropriate dimension at each level.
+   * This constructor uses a default initial depth of 0.
+   *
+   * @param dormRoomList the list of dorm rooms to build the KD-tree node from; must not be empty
+   * @throws IllegalArgumentException if {@code dormRoomList} is null or empty
+   */
+  public KDTreeNode(List<DormRoom> dormRoomList) throws IllegalArgumentException{
+    this(dormRoomList, 0);
+  }
+
+  private KDTreeNode(List<DormRoom> rooms, int depth) throws IllegalArgumentException {
+    // Checks that the dorm list is not empty
+    if (rooms == null || rooms.isEmpty()) {
+      throw new IllegalArgumentException(
+          "Cannot create a KDTreeNode from an empty or null list of rooms.");
+    }
+
+    // Determine the axis (dimension) to split on.
+    int axis = depth % K;
+
+    // Sort the rooms
+    rooms.sort(getComparatorForAxis(axis));
+
+    // Select the median room as the node's value.
+    int medianIndex = rooms.size() / 2;
+    this.value = rooms.get(medianIndex);
+
+    // Build left subtree (rooms before the median)
+    if (medianIndex > 0) { // Check that there are rooms left on the medium
+      List<DormRoom> leftRooms = rooms.subList(0, medianIndex);
+      this.left = new KDTreeNode(leftRooms, depth + 1);
+    } else {
+      this.left = null;
+    }
+
+    // Build right subtree (rooms after the median)
+    if (medianIndex < rooms.size() - 1) { // Check that there are rooms right on the medium
+      List<DormRoom> rightRooms = rooms.subList(medianIndex + 1, rooms.size());
+      this.right = new KDTreeNode(rightRooms, depth + 1);
+    } else {
+      this.right = null;
+    }
+  }
+
+
+  /**
+   * Creates and returns a comparator for sorting dorm rooms along a given axis.
+   * The axis mapping is:
+   * <ul>
+   *   <li>0: roomSize</li>
+   *   <li>1: roomCapacity</li>
+   *   <li>2: roomNumber</li>
+   * </ul>
+   *
+   * @param axis the axis to sort on, must be an integer between 0 and 2 inclusive
+   * @return a comparator that compares dorm rooms on the specified axis
+   * @throws IllegalArgumentException if the axis is out of range
+   */
+  private Comparator<DormRoom> getComparatorForAxis(int axis) {
```

Comment:
We should not be filtering by room number. We should be filtering by floor number and add floor number as a field to dorm room